File: internal/checks/sri_missing.go
package checks

import (
	"bytes"
	"context"
	"fmt"
	"net/url"
	"strings"

	"golang.org/x/net/html"

	"github.com/londonmax12/hyperz/internal/httpclient"
	"github.com/londonmax12/hyperz/internal/page"
	"github.com/londonmax12/hyperz/internal/scope"
)

// SRIMissing flags executable subresources loaded from a different origin that
// do not carry an integrity="sha…" attribute. Without SRI, an attacker who
// compromises the CDN or sits on the network path can swap the file for a
// malicious version and the browser will execute it.
//
// Scope is deliberately narrow:
//
//   - <script src=...> from a different host - SRI is the defined defense, no
//     SRI is high impact.
//   - <link rel="stylesheet|preload|modulepreload|prefetch" href=...> from a
//     different host - SRI is supported and applicable. Other link rels
//     (icon, canonical, dns-prefetch, manifest, alternate, ...) are excluded:
//     they either don't fetch executable content, don't validate via SRI in
//     any browser, or are navigation hints.
//
// <iframe> is intentionally not covered: there is no integrity mechanism for
// cross-origin iframe src in any current browser, so flagging it would be a
// guaranteed false positive.
type SRIMissing struct{}

func (SRIMissing) Name() string { return "sri-missing" }

func (SRIMissing) Level() Level { return LevelPassive }

// sriMissingBodyCap bounds how much HTML we tokenize. Matches the other
// passive HTML scanners; documents past this size lose late-page <script>
// coverage in exchange for predictable cost.
const sriMissingBodyCap = 2 << 20

// sriLinkRels is the set of <link rel> tokens for which SRI is meaningful.
// A rel attribute may carry multiple space-separated tokens; any match here
// qualifies the link for an SRI check.
var sriLinkRels = map[string]struct{}{
	"stylesheet":    {},
	"preload":       {},
	"modulepreload": {},
	"prefetch":      {},
}

func (c SRIMissing) Run(ctx context.Context, client *httpclient.Client, _ *scope.Scope, p page.Page) ([]Finding, error) {
	snap, err := ensureResponse(ctx, client, p, sriMissingBodyCap)
	if err != nil {
		return nil, err
	}
	if !isHTMLContentType(snap.Headers.Get("Content-Type")) {
		return nil, nil
	}
	if len(snap.Body) == 0 {
		return nil, nil
	}

	pageURL, err := url.Parse(p.URL)
	if err != nil || pageURL.Host == "" {
		return nil, nil
	}

	z := html.NewTokenizer(bytes.NewReader(snap.Body))
	var findings []Finding
	seen := map[string]struct{}{}

	for {
		tt := z.Next()
		if tt == html.ErrorToken {
			break
		}
		if tt != html.StartTagToken && tt != html.SelfClosingTagToken {
			continue
		}
		tag, hasAttr := z.TagName()
		if !hasAttr {
			continue
		}
		tagName := string(tag)

		var resourceAttr string
		switch tagName {
		case "script":
			resourceAttr = "src"
		case "link":
			resourceAttr = "href"
		default:
			continue
		}

		var resourceURL, integrityAttr, relAttr string
		for {
			key, val, more := z.TagAttr()
			switch strings.ToLower(string(key)) {
			case resourceAttr:
				resourceURL = string(val)
			case "integrity":
				integrityAttr = string(val)
			case "rel":
				relAttr = string(val)
			}
			if !more {
				break
			}
		}

		// <link> only fetches an SRI-eligible resource for specific rels.
		if tagName == "link" && !linkRelIsSRIEligible(relAttr) {
			continue
		}
		if strings.TrimSpace(resourceURL) == "" {
			continue
		}
		if integrityAttr != "" {
			continue
		}

		// Drop non-network schemes; they aren't subresources in any meaningful
		// sense and parsing them as URLs adds noise.
		lower := strings.ToLower(strings.TrimSpace(resourceURL))
		if strings.HasPrefix(lower, "data:") ||
			strings.HasPrefix(lower, "javascript:") ||
			strings.HasPrefix(lower, "blob:") ||
			strings.HasPrefix(lower, "#") {
			continue
		}

		ref, err := url.Parse(resourceURL)
		if err != nil {
			continue
		}
		// Resolve against the page so protocol-relative (//cdn/...) and bare
		// relative URLs land in the right host; matches what the browser does
		// when it goes to fetch the resource.
		resolved := pageURL.ResolveReference(ref)
		if resolved.Host == "" {
			continue
		}
		// Same-origin: the page can already substitute the file at will, SRI
		// adds nothing. Hostname() strips the port so default ports don't
		// look like a different host than an explicit one.
		if strings.EqualFold(resolved.Hostname(), pageURL.Hostname()) {
			continue
		}

		resolvedURL := resolved.String()
		dedupeKey := MakeKey(c.Name(), ScopeHost, p.URL, "url:"+resolvedURL)
		if _, dup := seen[dedupeKey]; dup {
			continue
		}
		seen[dedupeKey] = struct{}{}

		severity := SeverityMedium
		if tagName == "script" {
			severity = SeverityHigh
		}

		findings = append(findings, Finding{
			Check:    c.Name(),
			Target:   p.URL,
			URL:      p.URL,
			Severity: severity,
			Title:    fmt.Sprintf("cross-origin <%s> loaded without Subresource Integrity", tagName),
			Detail: fmt.Sprintf(
				"Page %s loads <%s> from %s without an integrity attribute. "+
					"An attacker who compromises that origin or sits on the network path between the "+
					"browser and the CDN can substitute the file with malicious content and the browser "+
					"will execute it as if it were trusted first-party code.",
				p.URL, tagName, resolvedURL,
			),
			CWE:   "CWE-345",
			OWASP: "A08:2021 Software and Data Integrity Failures",
			Remediation: "Add integrity=\"sha384-<base64-hash>\" (and crossorigin=\"anonymous\" for cross-origin loads) to the tag. " +
				"Most public CDNs (jsDelivr, unpkg, cdnjs) publish SRI hashes alongside their URLs; " +
				"for self-hosted assets generate one with `openssl dgst -sha384 -binary <file> | openssl base64 -A`. " +
				"Alternatively, host the file from the same origin so the integrity question collapses to TLS.",
			Evidence:  BuildEvidence("GET", p.URL, snap.Status, snap.Headers, ""),
			DedupeKey: dedupeKey,
		})
	}

	return findings, nil
}

// linkRelIsSRIEligible reports whether rel - which may carry multiple
// space-separated tokens - names a relationship that fetches a subresource
// SRI applies to. An empty rel falls through to false: a <link> without a
// rel is inert in browsers and not worth flagging.
func linkRelIsSRIEligible(rel string) bool {
	for _, tok := range strings.Fields(rel) {
		if _, ok := sriLinkRels[strings.ToLower(tok)]; ok {
			return true
		}
	}
	return false
}

File: internal/checks/html/sri_missing.lua
-- sri-missing: flags <script src> / <link rel="stylesheet|preload|
-- modulepreload|prefetch" href> loaded from a different host without
-- integrity=... Same-origin loads and <iframe> (which has no SRI
-- mechanism in any browser) are deliberately excluded.

local check = {
  name        = "sri-missing",
  level       = levels.passive,
  scope       = scopes.host,
  cwe         = "CWE-345",
  owasp       = "A08:2021 Software and Data Integrity Failures",
  remediation = 'Add integrity="sha384-<base64-hash>" (and crossorigin="anonymous" for cross-origin loads) to the tag. '
                .. "Most public CDNs (jsDelivr, unpkg, cdnjs) publish SRI hashes alongside their URLs; "
                .. "for self-hosted assets generate one with `openssl dgst -sha384 -binary <file> | openssl base64 -A`. "
                .. "Alternatively, host the file from the same origin so the integrity question collapses to TLS.",
  tier        = tiers.passive,
}

local LINK_RELS = {
  stylesheet    = true,
  preload       = true,
  modulepreload = true,
  prefetch      = true,
}

-- rel is a space-separated token list; any one of LINK_RELS qualifies.
-- Empty rel is inert in browsers and skipped.
local function link_rel_is_sri_eligible(rel)
  for tok in string.gmatch(rel, "%S+") do
    if LINK_RELS[string.lower(tok)] then return true end
  end
  return false
end

function check.run(ctx)
  local snap, err = ctx:ensure_response()
  if err then return nil, err end
  if not ctx.body.is_html_ct(snap.headers:get("Content-Type")) then
    return nil
  end
  if snap.body == "" then return nil end

  local page_url, perr = ctx.url.parse(ctx.page.url)
  if perr or not page_url or page_url.host == "" then return nil end

  local tags = ctx.html.iter_tags(snap.body, { "script", "link" })
  local findings = {}
  local seen = {}
  local evidence = ctx.evidence.build {
    method  = methods.get,
    url     = ctx.page.url,
    status  = snap.status,
    headers = snap.headers,
  }

  for _, tag in ipairs(tags) do
    local resource_attr
    if tag.tag == "script" then
      resource_attr = "src"
    elseif tag.tag == "link" then
      resource_attr = "href"
    end
    if resource_attr then
      local rel = tag.attr["rel"] or ""
      local resource_url = tag.attr[resource_attr] or ""
      local integrity = tag.attr["integrity"] or ""

      -- <link> only fetches an SRI-eligible resource for specific
      -- rels; everything else (icon, canonical, dns-prefetch, ...)
      -- has no SRI either-way.
      local eligible = true
      if tag.tag == "link" and not link_rel_is_sri_eligible(rel) then
        eligible = false
      end

      if eligible and integrity == "" and (resource_url:gsub("^%s+", ""):gsub("%s+$", "")) ~= "" then
        -- ctx.html.resolve_ref handles the non-network skip set
        -- (data:, javascript:, blob:, fragment).
        local resolved, ok = ctx.html.resolve_ref(ctx.page.url, resource_url)
        if ok then
          local ru, rperr = ctx.url.parse(resolved)
          if not rperr and ru and ru.host ~= "" then
            -- Same-origin: SRI adds nothing. hostname() strips port
            -- so default-port mismatches don't false-positive.
            if string.lower(ru.hostname) ~= string.lower(page_url.hostname) then
              local key = "url:" .. resolved
              if not seen[key] then
                seen[key] = true
                local sev_key = "medium"
                if tag.tag == "script" then sev_key = "high" end
                findings[#findings + 1] = {
                  severity = severity[sev_key],
                  title    = string.format("cross-origin <%s> loaded without Subresource Integrity", tag.tag),
                  detail   = string.format(
                    "Page %s loads <%s> from %s without an integrity attribute. "
                      .. "An attacker who compromises that origin or sits on the network path between the "
                      .. "browser and the CDN can substitute the file with malicious content and the browser "
                      .. "will execute it as if it were trusted first-party code.",
                    ctx.page.url, tag.tag, resolved),
                  evidence = evidence,
                  dedupe_parts = { key },
                }
              end
            end
          end
        end
      end
    end
  end

  return findings
end

return check
